### ISSUE TITLE
Precomputed specs to allow easy creation of new scales

### ISSUE THREAD

**Issue by supersergiy** [ZettaAI/zetta_utils#500]
* A list of scales to be created would be provided as a separate argument
* Optional reference scale would default to MIP0. (@trivoldus28 I was wrong before, having a reference scale would be useful in many scenarios)
* `chunk_size`, `dataset_size`, and `voxel_offset` maps could be used to specify small deviations from the reference scale

Questions:
* What if both new scale list and `info_field_overrides: {'scales': ...}` are provided? Should an exception be thrown, or should the new scales be created on top of the initial scales provided by the overrides?

**Comment by trivoldus28** [comment]
Thanks for making the issue @supersergiy . I'm gonna write up some draft examples of how the interface is going to be like and yea I think having a ref scale would be useful. I'll take a look of how `info_field_overrides` are being used first tho to cover most common use cases.

On handling `info_field_overrides` with `scales` specified, I think we should just throw an exception for now unless it's useful to merge it and adding scales.

**Comment by nkemnitz** [comment]
`size`, `voxel_offset`, `key` are the keys that are directly linked to resolution. At least during our current alignment and segmentations steps, the shrink factor is always `[2,2,1]`. For the final render step, we sometimes downsample in z as well to keep voxel size isotropic. `chunk_sizes` often stays the same across scales.

Ideally we would check that for all the requested scales, the reference values make sense.
E.g. requesting 3 scales with a voxel_offset of 30 at scale 0 will cause issues with the visualization at scale 3, because the image suddenly jumps half a pixel

**Comment by trivoldus28** [comment]
Example of what I think could work:
```

{
    "@type": "build_cv_layer"
    path: #DST_PATH
    ref_info: ...

    // add_scales_ref can either be a dictionary
    add_scales_ref: {
        "key": "4_4_40",
        "resolution": [4, 4, 40],
        "chunk_sizes": [32, 32, 1],
        "encoding": "raw",
        "size": #OUTPUT_SIZE,
        "voxel_offset": [0, 0, 0]
    }
    // Or an existing key from `ref_info` from which the dict will be copied from
    // Throws an error if `ref_info` is not provided
    add_scales_ref: "4_4_40"

    // Example adding 3 scales
    add_scales: [
        {
            // full specs for this scale
            "key": "8_8_80",
            "resolution": [8, 8, 40],
            "chunk_sizes": [32, 32, 1],
            "encoding": "raw",
            "size": #OUTPUT_SIZE,
            "voxel_offset": [1, 2, 3]
        },

        {   // Fill in missing parameters using defaults. We check info_xxx_size/offset/_map for values.
            // We get values from `add_scales_ref` (with `size` & `offset` scaled proportionally to resolution) if avail.
            // Error cases:
            // 1. No values are available.
            // 2. Conflicting values.
            // 3. Values are not integer after scaling.
            // To resolve, just provide the errored parameters here.
            "key": "16_16_40",
            "resolution": [16, 16, 40],
        },  

        [32, 32, 40],  // shorthand, equivalent to `{"key": 8_8_40, "resolution": [8, 8, 40]}`
        ]
}

```

**Comment by trivoldus28** [comment]
@supersergiy Could also try setting the default scale to be mip0 but I wonder if it's better to not and let the user be more explicit. For one, theoretically there could be multiple keys with the same res right?

**Comment by trivoldus28** [comment]
I'm still thinking there could be pitfalls with defaulting the ref scale to mip0. For instance, if the user adds a higher res scale on the first run, the next run may not be consistent to the first run.

As a middle ground we can let the user specify `add_scales_ref: "mip0"`.

**Comment by trivoldus28** [comment]
I would also appreciate not having to write and support a function to determine what is MIP0 ;). Maybe `add_scales_ref: "mip0"` can be a reach goal for later after the first implementation.

**Comment by supersergiy** [comment]
Fixed via #511

**Comment by supersergiy** [comment]
@trivoldus28 if you include `Fixes #{Issue number` in the PR text, it will automatically close the issue as complete on PR merge [[reference](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)]
